File: bench/csv.py
# The Python twin of csv.sl.
#
# `str.split` and `int` are the builtins, and `len` counts code points -- the same unit slate counts.
# The standard library's `csv` module would be the idiomatic reader for a real file and would not be
# the same work: it handles quoting and would be measuring a different program.


def build(rows):
    return "\n".join(f"{i},{i * 2},name{i % 100}" for i in range(rows))


def parse(text):
    total = 0

    for line in text.split("\n"):
        parts = line.split(",")

        total = total + int(parts[0]) + int(parts[1]) + len(parts[2])

    return total


def run():
    text = build(20000)
    total, turns = 0, 0

    while turns < 30:
        total = total + parse(text)
        turns = turns + 1

    return total


print(run())

File: bench/csv.rb
# The Ruby twin of csv.sl.
#
# `String#split` and `to_i` are the builtins, and `length` counts characters -- slate's unit, and
# read without a scan, this text being ASCII. The standard library's `CSV` would be the idiomatic
# reader for a real file and would not be the same work: it handles quoting.

def build(rows)
  lines = []
  i = 0

  while i < rows
    lines << "#{i},#{i * 2},name#{i % 100}"
    i = i + 1
  end

  lines.join("\n")
end

def parse(text)
  total = 0

  text.split("\n").each do |line|
    parts = line.split(",")

    total = total + parts[0].to_i + parts[1].to_i + parts[2].length
  end

  total
end

def run
  text = build(20_000)
  total = 0
  turns = 0

  while turns < 30
    total = total + parse(text)
    turns = turns + 1
  end

  total
end

puts run
